### DISCUSSION TITLE
Are there any Nuxt3 features that are not available with the async function?

### DISCUSSION THREAD

**Discussion by mogi86** [nuxt/nuxt#16933] [Questions]
Hi. I sometimes face to error when I use `useCookie()` and `useRuntimeConfig()` in async function .
Shouldn't these features be used within asynchronous function?

**Comment by danielroe** [comment]
See https://github.com/nuxt/framework/issues/5740#issuecomment-1229197529 for an explanation of what's going on.

**Comment by mogi86** [comment]
Thank you.
So, I need to use `<script setup>`?

**Reply by k0mar12** [reply]
@mogi86 No. You cant use composable function inside another functions.

**Reply by mogi86** [reply]
Thank you to reply.

What I want to do .
- Call an external API and receive a response.
- Store the response of the API into a cookie.
- I want to execute this process on all pages.

How should I implement this? As below source code, do I need to put all the processing in `setup()` without using another functions If I want to use `composable function`? Actually, there is more code because of the business logic. Therefore, I moved the processing to another functions.
Sorry, my understanding may be wrong.

```ts
<script setup lang="ts">

const { data } = await useFetch(`${useRuntimeConfig().public.domain}/api/something`)
useCookie<any>('response').value = data.value

</script>
```

**Reply by k0mar12** [reply]
@mogi86 oh, i got. I had a similar problem.

I have created a composable helper **useSsrHelper**

```
import { appendHeader } from 'h3'
import { parse, splitCookiesString as split } from 'set-cookie-parser'
import { serialize } from 'cookie'

export const useSsrHelper = () => {
  if (process.client) {
    return {
      options: () => ({
        Accept: 'application/json'
      }),
      host: window.location.hostname,
      referer: window.document.referrer
    }
  }

  /**
   * TODO: nuxtApp.ssrContext.req.headers.referer
   *
   * Not ready in nuxt 3, need to check new releases notes
   */
  const nuxtApp = useNuxtApp()
  const host = nuxtApp.ssrContext.req.headers['x-proto-host']
  const referer = `${host}/${nuxtApp.ssrContext.req.originalUrl}`
  const goodResponses = [200, 201, 202, 203, 204, 205, 206, 207, 208, 226]

  /**
     * Parse cookie from ajax response to add them to final response
     *
     * @param cookies
     * @returns
     */
  const parseCookie = (cookies) => {
    return (parse(split(cookies))).map(cookie => serialize(cookie.name, cookie.value, cookie))
  }

  /**
     * Default options for $fetch
     *
     * @returns
     */
  const options = () => {
    return {
      baseURL: useRuntimeConfig().APP_SERVER_URL,
      headers: {
        ...useRequestHeaders(['cookie']),
        Accept: 'application/json',
        referer
      },
      // eslint-disable-next-line
      async onResponse ({ response }) {
        const { status, headers } = response

        if (goodResponses.includes(status)) {
          appendHeader(useRequestEvent(nuxtApp), 'Set-Cookie', parseCookie(headers.get('set-cookie')))
        }
      }
    }
  }

  return {
    options, host, referer
  }
}
```
Also its needed two dependencies: **set-cookie-parser** and **cookie**

After that, u can set ur cookie from response

```
const { options } = useSsrHelper()

await useFetch('/url', options())
```

**Reply by k0mar12** [reply]
@mogi86 Another way not exist at this moment. useCookie will not work after async function

**Reply by mogi86** [reply]
@k0mar12 
Thank you for your kind reply! I understand.
